processor/src/main/java/org/jboss/pnc/processor/ClientGenerator.java

```diff
@@ -204,10 +143,107 @@ private boolean process0(Set<? extends TypeElement> annotations, RoundEnvironmen
+                        .addStatement("return Optional.empty()");
+            } else if (restApiMethod.getAnnotation(POST.class) != null) {
+                builder
+                        .returns(TypeName.get(returnGeneric))
+                        .addStatement("return getEndpoint()." + restApiMethod.getSimpleName() + "(" + parametersList + ").getContent()");
+            }
+        } else if (ClassName.get(returnType).toString().equals("java.lang.String")) {
+            //string response
+            addDefaultParameters(restApiMethod, builder, withParameters);
+            builder
+                    .returns(ParameterizedTypeName.get(ClassName.get(Optional.class), TypeName.get(String.class)))
+                    .addStatement("return Optional.ofNullable(getEndpoint()." + restApiMethod.getSimpleName() + "(" + parametersList + "))")
+                    .nextControlFlow("catch ($T e)", NotFoundException.class)
+                    .addStatement("return Optional.empty()");
+        } else if (ClassName.get(returnType).toString().equals("void")) {
+            //void response
+            addDefaultParameters(restApiMethod, builder, withParameters);
+            builder
+                    .addException(ClassName.get("org.jboss.pnc.client", "RemoteResourceNotFoundException"))
+                    .returns(TypeName.get(void.class))
+                    .addStatement("getEndpoint()." + restApiMethod.getSimpleName() + "(" + parametersList + ")")
+                    .nextControlFlow("catch ($T e)", NotFoundException.class)
+                    .addStatement("throw new RemoteResourceNotFoundException(e)");
+        } else {
+            //any other return types
+            addDefaultParameters(restApiMethod, builder, withParameters);
+            builder
+                    .addException(ClassName.get("org.jboss.pnc.client", "ClientException"))
+                    .returns(TypeName.get(returnType))
+                    .addStatement("return getEndpoint()." + restApiMethod.getSimpleName() + "(" + parametersList + ")")
+                    .nextControlFlow("catch ($T e)", NotFoundException.class)
+                    .addStatement("throw new RemoteResourceNotFoundException(e)");
+        }
+
+        return builder
+                .nextControlFlow("catch ($T e)", ClientErrorException.class)
+                .addStatement("throw new RemoteResourceException(e)")
+                .endControlFlow()
+                .build();
+    }
+
+    private void addDefaultParameters(ExecutableElement restApiMethod,
+                                      MethodSpec.Builder builder,
+                                      boolean withParameters) {
         for (VariableElement parameter : restApiMethod.getParameters()) {
             builder.addParameter(TypeName.get(parameter.asType()), parameter.getSimpleName().toString());
         }
+        if (withParameters) {
+            builder.addParameter(MAP_OF_STRINGS, "parameters");
```

Comment:
What about this way:
builder.addParameter(ParameterizedTypeName.get(Map.class, String.class, String.class), "parameters");

Reply:
I'd rather change the MAP_OF_STRINGS value to `ParametrizedTypeName.get...` since it's used twice and it's a lot of characters.
WDYT?